ballista/rust/scheduler/src/state/execution_graph_dot.rs

```diff
@@ -0,0 +1,537 @@
+                        &mut dot,
+                        &stage_name,
+                        &stage.plan,
+                        0,
+                    )?);
+                }
+                ExecutionStage::Resolved(stage) => {
+                    writeln!(&mut dot, "\t\tlabel = \"Stage {} [Resolved]\";", id)?;
+                    stage_meta.push(write_stage_plan(
+                        &mut dot,
+                        &stage_name,
+                        &stage.plan,
+                        0,
+                    )?);
+                }
+                ExecutionStage::Running(stage) => {
+                    writeln!(&mut dot, "\t\tlabel = \"Stage {} [Running]\";", id)?;
+                    stage_meta.push(write_stage_plan(
+                        &mut dot,
+                        &stage_name,
+                        &stage.plan,
+                        0,
+                    )?);
+                }
+                ExecutionStage::Completed(stage) => {
+                    writeln!(&mut dot, "\t\tlabel = \"Stage {} [Completed]\";", id)?;
+                    stage_meta.push(write_stage_plan(
+                        &mut dot,
+                        &stage_name,
+                        &stage.plan,
+                        0,
+                    )?);
+                }
+                ExecutionStage::Failed(stage) => {
+                    writeln!(&mut dot, "\t\tlabel = \"Stage {} [FAILED]\";", id)?;
+                    stage_meta.push(write_stage_plan(
+                        &mut dot,
+                        &stage_name,
+                        &stage.plan,
+                        0,
+                    )?);
+                }
+            }
+            cluster += 1;
+            writeln!(&mut dot, "\t}}")?; // end of subgraph
+        }
+
+        // write links between stages
+        for meta in &stage_meta {
+            let mut links = vec![];
+            for (reader_node, parent_stage_id) in &meta.readers {
+                // shuffle write node is always node zero
+                let parent_shuffle_write_node = format!("stage_{}_0", parent_stage_id);
+                links.push(format!("{} -> {}", parent_shuffle_write_node, reader_node,));
+            }
+            // keep the order deterministic
+            links.sort();
+            for link in links {
+                writeln!(&mut dot, "\t{}", link)?;
+            }
+        }
+
+        writeln!(&mut dot, "}}")?; // end of digraph
+
+        Ok(dot)
+    }
+}
+
+/// Write the query tree for a single stage and build metadata needed to later draw
+/// the links between the stages
+fn write_stage_plan(
+    f: &mut String,
+    prefix: &str,
+    plan: &Arc<dyn ExecutionPlan>,
+    i: usize,
+) -> Result<StagePlanState, fmt::Error> {
+    let mut state = StagePlanState {
+        readers: HashMap::new(),
+    };
+    write_plan_recursive(f, prefix, plan, i, &mut state)?;
+    Ok(state)
+}
+
+fn write_plan_recursive(
```

Comment:
We could use/move something similar to DataFusion (minus the stage part, currently).